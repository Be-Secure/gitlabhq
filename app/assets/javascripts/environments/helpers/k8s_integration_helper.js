import { CLUSTER_AGENT_ERROR_MESSAGES } from '../constants';

export function humanizeClusterErrors(reason) {
  const errorReason = String(reason).toLowerCase();
  const errorMessage = CLUSTER_AGENT_ERROR_MESSAGES[errorReason];
  return errorMessage || CLUSTER_AGENT_ERROR_MESSAGES.other;
}
