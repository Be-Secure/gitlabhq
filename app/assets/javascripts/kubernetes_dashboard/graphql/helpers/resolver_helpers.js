import { CoreV1Api, Configuration, WatchApi, EVENT_DATA } from '@gitlab/cluster-client';

export const handleClusterError = async (err) => {
  if (!err.response) {
    throw err;
  }

  const errorData = await err.response.json();
  throw errorData;
};

export const buildWatchPath = ({ resource, api = 'api/v1', namespace = '' }) => {
  return namespace ? `/${api}/namespaces/${namespace}/${resource}` : `/${api}/${resource}`;
};

export const mapWorkloadItem = (item) => {
  if (item.metadata) {
    const metadata = {
      ...item.metadata,
      annotations: item.metadata?.annotations || {},
      labels: item.metadata?.labels || {},
    };
    return { status: item.status, metadata };
  }
  return { status: item.status };
};

export const mapSetItem = (item) => {
  const status = {
    ...item.status,
    readyReplicas: item.status?.readyReplicas || null,
  };

  const metadata =
    {
      ...item.metadata,
      annotations: item.metadata?.annotations || {},
      labels: item.metadata?.labels || {},
    } || null;

  const spec = item.spec || null;

  return { status, metadata, spec };
};

export const mapJobItem = (item) => {
  const metadata = {
    ...item.metadata,
    annotations: item.metadata?.annotations || {},
    labels: item.metadata?.labels || {},
  };

  const status = {
    failed: item.status?.failed || 0,
    succeeded: item.status?.succeeded || 0,
  };

  return {
    status,
    metadata,
    spec: item.spec,
  };
};

export const mapServicesItems = (item) => {
  const { type, clusterIP, externalIP, ports } = item.spec;

  return {
    metadata: {
      ...item.metadata,
      annotations: item.metadata?.annotations || {},
      labels: item.metadata?.labels || {},
    },
    spec: {
      type,
      clusterIP: clusterIP || '-',
      externalIP: externalIP || '-',
      ports,
    },
  };
};

export const mapCronJobItem = (item) => {
  const metadata = {
    ...item.metadata,
    annotations: item.metadata?.annotations || {},
    labels: item.metadata?.labels || {},
  };

  const status = {
    active: item.status?.active || 0,
    lastScheduleTime: item.status?.lastScheduleTime || null,
  };

  return {
    status,
    metadata,
    spec: item.spec,
  };
};

export const watchWorkloadItems = ({
  client,
  query,
  configuration,
  namespace,
  watchPath,
  queryField,
  mapFn = mapWorkloadItem,
}) => {
  const config = new Configuration(configuration);
  const watcherApi = new WatchApi(config);

  watcherApi
    .subscribeToStream(watchPath, { watch: true })
    .then((watcher) => {
      let result = [];

      watcher.on(EVENT_DATA, (data) => {
        result = data.map(mapFn);

        client.writeQuery({
          query,
          variables: { configuration, namespace },
          data: { [queryField]: result },
        });
      });
    })
    .catch((err) => {
      handleClusterError(err);
    });
};

export const getK8sPods = ({
  client,
  query,
  configuration,
  namespace = '',
  enableWatch = false,
}) => {
  const config = new Configuration(configuration);

  const coreV1Api = new CoreV1Api(config);
  const podsApi = namespace
    ? coreV1Api.listCoreV1NamespacedPod({ namespace })
    : coreV1Api.listCoreV1PodForAllNamespaces();

  return podsApi
    .then((res) => {
      if (enableWatch) {
        const watchPath = buildWatchPath({ resource: 'pods', namespace });
        watchWorkloadItems({
          client,
          query,
          configuration,
          namespace,
          watchPath,
          queryField: 'k8sPods',
        });
      }

      const data = res?.items || [];
      return data.map(mapWorkloadItem);
    })
    .catch(async (err) => {
      try {
        await handleClusterError(err);
      } catch (error) {
        throw new Error(error.message);
      }
    });
};
