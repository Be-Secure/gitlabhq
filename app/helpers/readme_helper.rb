# frozen_string_literal: true

module ReadmeHelper
  # @return [Hash]
  def vue_readme_header_additional_data
    {}
  end
end

ReadmeHelper.prepend_mod_with("ReadmeHelper")
