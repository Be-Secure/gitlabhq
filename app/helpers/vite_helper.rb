# frozen_string_literal: true

module ViteHelper
  def vite_enabled?
    # vite is not production ready yet
    return false if Rails.env.production?
    # Enable vite if explicitly turned on in the GDK
    return Gitlab::Utils.to_boolean(ViteRuby.env['VITE_ENABLED'], default: false) if ViteRuby.env.key?('VITE_ENABLED')

    # Enable vite the legacy way (in case GDK hasn't been updated)
    # This is going to be removed with https://gitlab.com/gitlab-org/gitlab/-/issues/431041
    Rails.env.development? ? Feature.enabled?(:vite) : false
  end
end
