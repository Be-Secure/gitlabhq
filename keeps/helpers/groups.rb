# frozen_string_literal: true

module Keeps
  module Helpers
    class Groups
      Error = Class.new(StandardError)

      def group_for_feature_category(category)
        @groups ||= {}
        @groups[category] ||= fetch_groups.find do |_, group|
          group['categories'].present? && group['categories'].include?(category)
        end&.last
      end

      def pick_reviewer(group, identifiers)
        random_engineer = Digest::SHA256.hexdigest(identifiers.join).to_i(16) % group['backend_engineers'].size

        group['backend_engineers'][random_engineer]
      end

      private

      def fetch_groups
        @groups_json ||= begin
          response = Gitlab::HTTP.get('https://about.gitlab.com/groups.json')

          unless (200..299).cover?(response.code)
            raise Error,
              "Failed to get group information with response code: #{response.code} and body:\n#{response.body}"
          end

          Gitlab::Json.parse(response.body)
        end
      end
    end
  end
end
