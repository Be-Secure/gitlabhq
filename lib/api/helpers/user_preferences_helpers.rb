# frozen_string_literal: true

module API
  module Helpers
    module UserPreferencesHelpers
      extend ActiveSupport::Concern
      extend Grape::API::Helpers

      def update_user_namespace_settings(attrs)
        # This method will be redefined in EE.
        attrs
      end
    end
  end
end

API::Helpers::UserPreferencesHelpers.prepend_mod_with('API::Helpers::UserPreferencesHelpers')
