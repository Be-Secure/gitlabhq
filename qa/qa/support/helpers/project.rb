# frozen_string_literal: true

module QA
  module Support
    module Helpers
      module Project
        def wait_until_project_is_ready(project)
          # Repository can take a short time to become ready after project is created
          Support::Retrier.retry_on_exception(sleep_interval: 5) do
            create(:commit, project: project, commit_message: 'Add new file', actions: [
              { action: 'create', file_path: 'new_file', content: '# This is a new file' }
            ])
          end
        end
      end
    end
  end
end
