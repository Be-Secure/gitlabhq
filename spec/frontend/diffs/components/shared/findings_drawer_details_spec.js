import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import FindingsDrawerDetails from '~/diffs/components/shared/findings_drawer_details.vue';
import {
  mockFindingDetected,
  mockFindingsMultiple,
  mockProject,
} from '../../mock_data/findings_drawer';

describe('Findings Drawer Details', () => {
  let wrapper;

  const findingDetailsProps = {
    drawer: mockFindingDetected,
    project: mockProject,
  };

  const createWrapper = (findingDetailsOverrides = {}) => {
    const propsData = {
      drawer: findingDetailsProps.drawer,
      project: findingDetailsProps.project,
      ...findingDetailsOverrides,
    };

    wrapper = shallowMountExtended(FindingsDrawerDetails, {
      propsData,
    });
  };

  const findTitle = () => wrapper.findByTestId('findings-drawer-title');

  describe('General Rendering', () => {
    it('renders without errors', () => {
      createWrapper();
      expect(wrapper.exists()).toBe(true);
    });

    it('matches the snapshot with dismissed badge', () => {
      createWrapper();
      expect(wrapper.element).toMatchSnapshot();
    });

    it('matches the snapshot with detected badge', () => {
      createWrapper();
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Active Index Handling', () => {
    it('watcher sets active index on drawer prop change', () => {
      createWrapper({ drawer: mockFindingsMultiple[2] });
      expect(findTitle().props().value).toBe(mockFindingsMultiple[2].title);
    });
  });
});
