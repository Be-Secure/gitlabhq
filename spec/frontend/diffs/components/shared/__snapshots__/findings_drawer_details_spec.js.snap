// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Findings Drawer Details General Rendering matches the snapshot with detected badge 1`] = `
<ul
  class="!gl-pb-0 gl-border-b-initial gl-list-none gl-mb-0"
>
  <drawer-item-stub
    data-testid="findings-drawer-title"
    description="Name"
    value="mockedtitle"
  />
  <drawer-item-stub
    description="Status"
    value=""
  >
    <gl-badge-stub
      class="text-capitalize"
      iconsize="md"
      tag="span"
      target="_self"
      variant="warning"
    >
      detected
    </gl-badge-stub>
  </drawer-item-stub>
  <drawer-item-stub
    description="Description"
    value="fakedesc"
  />
  <drawer-item-stub
    description="Project"
    value=""
  >
    <gl-link-stub
      href="/testpath"
    >
      testname
    </gl-link-stub>
  </drawer-item-stub>
  <drawer-item-stub
    description="File"
    value=""
  />
  <drawer-item-stub
    description="Identifiers"
    value=""
  >
    <span>
      <gl-link-stub
        href="https://semgrep.dev/r/gitlab.eslint.detect-disable-mustache-escape"
      >
        eslint.detect-disable-mustache-escape
      </gl-link-stub>
    </span>
  </drawer-item-stub>
  <drawer-item-stub
    description="Tool"
    value="SAST"
  />
  <drawer-item-stub
    description="Engine"
    value="testengine name"
  />
</ul>
`;

exports[`Findings Drawer Details General Rendering matches the snapshot with dismissed badge 1`] = `
<ul
  class="!gl-pb-0 gl-border-b-initial gl-list-none gl-mb-0"
>
  <drawer-item-stub
    data-testid="findings-drawer-title"
    description="Name"
    value="mockedtitle"
  />
  <drawer-item-stub
    description="Status"
    value=""
  >
    <gl-badge-stub
      class="text-capitalize"
      iconsize="md"
      tag="span"
      target="_self"
      variant="warning"
    >
      detected
    </gl-badge-stub>
  </drawer-item-stub>
  <drawer-item-stub
    description="Description"
    value="fakedesc"
  />
  <drawer-item-stub
    description="Project"
    value=""
  >
    <gl-link-stub
      href="/testpath"
    >
      testname
    </gl-link-stub>
  </drawer-item-stub>
  <drawer-item-stub
    description="File"
    value=""
  />
  <drawer-item-stub
    description="Identifiers"
    value=""
  >
    <span>
      <gl-link-stub
        href="https://semgrep.dev/r/gitlab.eslint.detect-disable-mustache-escape"
      >
        eslint.detect-disable-mustache-escape
      </gl-link-stub>
    </span>
  </drawer-item-stub>
  <drawer-item-stub
    description="Tool"
    value="SAST"
  />
  <drawer-item-stub
    description="Engine"
    value="testengine name"
  />
</ul>
`;
